pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+        TPBs.
+    """
+    g = construct_tpb_graph(experiments)
+    _, cliqs = clique_removal(g)
+    new_cliqs = []
+    for cliq in cliqs:
+        new_cliq = []
+        for exp in cliq:
+            # duplicate `count` times
+            new_cliq += [exp] * g.nodes[exp]['count']
+
+        new_cliqs += [new_cliq]
+
+    return ExperimentSuite(new_cliqs, program=experiments.program, qubits=experiments.qubits)
+
+
+@dataclass(frozen=True)
+class ExperimentResult:
+    experiment: Experiment
+    expectation: Union[float, complex]
+    stddev: float
+
+    def __str__(self):
+        return f'{self.experiment}: {self.expectation} +- {self.stddev}'
+
+    def __repr__(self):
+        return f'ExperimentResult[{self}]'
+
+    def serializable(self):
+        return {
+            'type': 'ExperimentResult',
+            'experiment': self.experiment,
+            'expectation': self.expectation,
+            'stddev': self.stddev,
+        }
+
+
+def measure_observables(qc: QuantumComputer, experiment_suite: ExperimentSuite, n_shots=1000,
+                        progress_callback=None, active_reset=False):
+    """
+    Measure all the observables in an ExperimentSuite.
+
+    :param qc:
+    :param experiment_suite:
+    :param n_shots:
+    """
+    for i, experiments in enumerate(experiment_suite):
+        log.info(f"Collecting bitstrings for the {len(experiments)} experiments: {experiments}")
+
+        total_prog = Program()
+        already_prepped = dict()
+        for exp in experiments:  # todo: find exp with max weight (for in_operator)
+            for idx, op_str in exp.in_operator:
+                if idx in already_prepped:
+                    assert already_prepped[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_prep(op_str, idx)
+                    already_prepped[idx] = op_str
+        total_prog += experiment_suite.program
+
+        already_meased = dict()
+        for exp in experiments:  # todo: find exp with max weight (for out_operator)
+            for idx, op_str in exp.out_operator:
+                if idx in already_meased:
+                    assert already_meased[idx] == op_str
+                else:
+                    total_prog += _local_pauli_eig_meas(op_str, idx)
+                    already_meased[idx] = op_str
+        bitstrings = qc.run_and_measure(total_prog, n_shots, active_reset=active_reset)
+        obs_strings = {q: 1 - 2 * bitstrings[q] for q in bitstrings}
+
+        if progress_callback is not None:
+            progress_callback(i, len(experiment_suite))
+
+        for exp in experiments:
+            measured_qubits = []
+            for idx, op_str in exp.out_operator:
+                measured_qubits.append(idx)
+            measured_qubits = np.array(measured_qubits)
+            log.debug(f"Considering {exp}... 'measuring' qubits {measured_qubits}")
+            assert exp.in_operator.coefficient == 1
```

Comment:
What’s the reason for this? This prevents us from obtaining ExperimentResults for ExperimentSuites with Experiments with coefficients unequal to 1, e.g. the following ExperimentSuite is perfectly acceptable
```
import pyquil.operator_estimation as oe
import pyquil.paulis as pl
from pyquil import get_qc, Program

expt1 = oe.Experiment(pl.sX(1) * pl.sZ(0), pl.sX(2) * pl.sY(1) * pl.sZ(0))
expt2 = oe.Experiment(pl.sZ(1) * pl.sX(0), pl.sY(2) * pl.sZ(1) * pl.sX(0))
expt3 = oe.Experiment(pl.sX(2) * pl.sY(1) * pl.sZ(0), pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt4 = oe.Experiment(-1.3 * pl.sX(2) * pl.sY(1) * pl.sZ(0), 1.5 * pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt_suite = oe.ExperimentSuite([expt1, expt2, expt3, expt4], Program(H(0), H(1), H(2)), [0, 1, 2])
```
But if we try to run measure_observables(qc, expt_suite), we’d get an error with the last experiment.
```
qc = get_qc('9q-square-qvm')
for expt_result in oe.measure_observables(qc, expt_suite):
    print (expt_result)
```
we'd get the output
```
1*X1Z0→1*X2Y1Z0: 0.038000000000000034 +- 0.0315841485805529
1*Z1X0→1*Y2Z1X0: -0.052000000000000046 +- 0.0315642155026812
1*X2Y1Z0→1*Z2X1Y0: -0.02400000000000002 +- 0.031597872941474626
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-282-ac4fed168edd> in <module>
      1 qc = get_qc('9q-square-qvm')
----> 2 for expt_result in oe.measure_observables(qc, expt_suite):
      3     print (expt_result)

~/Documents/pyquil_mpharrigan_forked/pyquil/pyquil/operator_estimation.py in measure_observables(qc, experiment_suite, n_shots, progress_callback, active_reset)
    393 
    394         for exp in experiments:
--> 395             measured_qubits = []
    396             for idx, op_str in exp.out_operator:
    397                 measured_qubits.append(idx)

AssertionError: 
```

Reply:
this is for the input operator which shouldn't have a coefficient. It isn't really an operator ... it's a specification of the starting state. Maybe we should rename/rework this part

Reply:
updated assert message

Reply:
We should include a similar assert statement (and message) in the `Experiment` class, so the user already knows something is wrong when they write the following code:
```
import pyquil.operator_estimation as oe
import pyquil.paulis as pl
from pyquil import get_qc, Program
from pyquil.gates import H

expt1 = oe.Experiment(pl.sX(1) * pl.sZ(0), pl.sX(2) * pl.sY(1) * pl.sZ(0))
expt2 = oe.Experiment(pl.sZ(1) * pl.sX(0), pl.sY(2) * pl.sZ(1) * pl.sX(0))
expt3 = oe.Experiment(pl.sX(2) * pl.sY(1) * pl.sZ(0), pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt4 = oe.Experiment(-1.3 * pl.sX(2) * pl.sY(1) * pl.sZ(0), 1.5 * pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt_suite = oe.ExperimentSuite([expt1, expt2, expt3, expt4], Program(H(0), H(1), H(2)), [0, 1, 2])
```
Ideally, the definition of `expt4` above should throw an error. Currently, the code above runs without complaint, and the user wouldn't know there's a problem until running something like:
```
qc = get_qc('9q-square-qvm')
for expt_result in oe.measure_observables(qc, expt_suite):
    print (expt_result)
```
in which case they'd run into the `AssertionError`.